wgengine/magicsock/relaymanager.go

```diff
@@ -17,35 +25,150 @@ import (
-	h.mu.Lock()
-	defer h.mu.Unlock()
-	h.initLocked()
+// cancelOutstandingWork cancels any in-progress work for 'ep'.
+func (r *relayManager) cancelOutstandingWork(ep *endpoint) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	existing, ok := r.relaySetupWorkByEndpoint[ep]
+	if ok {
+		existing.cancel()
+		existing.wg.Wait()
+		delete(r.relaySetupWorkByEndpoint, ep)
+	}
+}
+
+// allocateAndHandshakeAllServers kicks off allocation and handshaking of relay
+// endpoints for 'ep' on all known relay servers, canceling any existing
+// in-progress work.
+func (r *relayManager) allocateAndHandshakeAllServers(ep *endpoint) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.initLocked()
+	existing, ok := r.relaySetupWorkByEndpoint[ep]
+	if ok {
+		existing.cancel()
+		existing.wg.Wait()
+		delete(r.relaySetupWorkByEndpoint, ep)
+	}
+	if len(r.serversByAddrPort) == 0 {
+		return
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	started := &relaySetupWork{ep: ep, cancel: cancel, wg: &sync.WaitGroup{}}
+	for k := range r.serversByAddrPort {
+		started.wg.Add(1)
+		go r.allocateAndHandshakeForServer(ctx, started.wg, ep, k)
+	}
+	r.relaySetupWorkByEndpoint[ep] = started
+
+	go func() {
+		started.wg.Wait()
+		started.cancel()
+		r.mu.Lock()
+		defer r.mu.Unlock()
+		maybeCleanup, ok := r.relaySetupWorkByEndpoint[ep]
+		if ok && maybeCleanup == started {
+			// A subsequent call to allocateAndHandshakeAllServers may have raced to
+			// delete the associated key/value, so ensure the work we are
+			// cleaning up from the map is the same as the one we were waiting
+			// to finish.
+			delete(r.relaySetupWorkByEndpoint, ep)
+		}
+	}()
+}
+
+func (r *relayManager) handleNewServerEndpoint(ctx context.Context, wg *sync.WaitGroup, server netip.AddrPort, se udprelay.ServerEndpoint) {
 	// TODO(jwhited): implement
 }
+
+func (r *relayManager) allocateAndHandshakeForServer(ctx context.Context, wg *sync.WaitGroup, ep *endpoint, server netip.AddrPort) {
+	defer wg.Done()
+	var b bytes.Buffer
+	remoteDisco := ep.disco.Load()
+	if remoteDisco == nil {
+		return
+	}
+	type allocateRelayEndpointReq struct {
+		DiscoKeys []key.DiscoPublic
+	}
+	a := &allocateRelayEndpointReq{
+		DiscoKeys: []key.DiscoPublic{ep.c.discoPublic, remoteDisco.key},
+	}
+	err := json.NewEncoder(&b).Encode(a)
+	if err != nil {
+		return
+	}
+	const reqTimeout = time.Second * 10
+	reqCtx, cancel := context.WithTimeout(ctx, reqTimeout)
+	defer cancel()
+	req, err := http.NewRequestWithContext(reqCtx, httpm.POST, "http://"+server.String()+"/relay/endpoint", &b)
+	if err != nil {
+		return
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return
+	}
+	var se udprelay.ServerEndpoint
+	err = json.NewDecoder(io.LimitReader(resp.Body, 4096)).Decode(&se)
+	if err != nil {
+		return
+	}
+	r.handleNewServerEndpoint(ctx, wg, server, se)
```

Comment:
:question: Question: What would `handleNewServerEndpoint` do with the `wg` WaitGroup?

Reply:
`udprelay.ServerEndpoint.AddrPorts` contains all the candidate ip:port pairs the relay server could be reachable over. `handleNewServerEndpoint` will be responsible for starting a 3-way handshake with those addr ports which may involve more goroutines and positive delta to `wg.Add()`.